include/upstream_temperature_feedback.hxx

```diff
@@ -0,0 +1,178 @@
+/// between the upstream temperature at y=0 and the specified value
+struct UpstreamTemperatureFeedback : public Component {
+
+  /// Inputs
+  ///  - <name> (e.g. "d+")
+  ///    - temperature_upstream        Upstream temperature (y=0) in eV
+  ///    - temperature_controller_p    Feedback proportional to error
+  ///    - temperature_controller_i    Feedback proportional to error integral
+  ///    - temperature_integral_positive  Force integral term to be positive? (default: false)
+  ///    - temperature_source_positive    Force temperature source to be positive? (default: true)
+  ///    - diagnose           Output diagnostic information?
+  ///
+  ///  - N<name>  (e.g. "Nd+")
+  ///    - source_shape  The initial source that is scaled by a time-varying factor
+  ///
+  UpstreamTemperatureFeedback(std::string name, Options& alloptions, Solver*) : name(name) {
+
+    Options& options = alloptions[name];
+    const auto& units = alloptions["units"];
+    BoutReal Tnorm = get<BoutReal>(units["eV"]);
+    BoutReal Nnorm = get<BoutReal>(units["inv_meters_cubed"]);
+    BoutReal Omega_ci = 1. / units["seconds"].as<BoutReal>();
+
+    BoutReal Pnorm = SI::qe * Tnorm * Nnorm; // Pressure normalisation
+    BoutReal SPnorm = Pnorm * Omega_ci; // Pressure-source normalisation [Pa/s] or [W/m^3] if converted to energy
+
+    temperature_upstream =
+        options["temperature_upstream"].doc("Upstream temperature (at y=0) [eV]").as<BoutReal>()
+        / Tnorm;
+
+    temperature_controller_p = options["temperature_controller_p"]
+                               .doc("Feedback controller proportional (p) parameter")
+                               .withDefault(1e-2);
+    temperature_controller_i = options["temperature_controller_i"]
+                               .doc("Feedback controller integral (i) parameter")
+                               .withDefault(1e-3);
+
+    temperature_integral_positive = options["temperature_integral_positive"]
+                                    .doc("Force integral term to be positive?")
+                                    .withDefault<bool>(false);
+    temperature_source_positive = options["temperature_source_positive"]
+                                  .doc("Force source to be positive?")
+                                  .withDefault<bool>(true);
+
+    // NOTE: temperature_error_integral should be set from the restart file here.
+    // There doesn't seem to be a way to get the restart Options,
+    // so for now this is set in restartVars
+
+    // Source shape the same as used in EvolvePressure
+    pressure_source_shape =
+      alloptions[std::string("P") + name]["source_shape"]
+            .doc("Source term in ddt(P" + name + std::string("). Units [Pa/s], note P = 2/3 E."))
+            .withDefault(Field3D(0.0))
+        / (SPnorm);
+
+    diagnose = options["diagnose"]
+                   .doc("Output additional diagnostics?")
+                   .withDefault<bool>(false);
```

Comment:
Something that I can't work out how to fix: with these changes the temperatures are no longer written into the output file by default. I need to add `diagnose = true` to each of the species (`e`, `d`, `d+`). This is the case even if I'm not using the temperature controller. I could change this line to `true` which would presumably fix the problem, but I'd prefer to understand why it's occuring.
@bendudson @mikekryjak thoughts?

Reply:
The `diagnose` flag has been set up on a per-species basis throughout the code, so this is expected behaviour. The `options` on line 64 is the options for the species and picks up settings immediately below the species header, e.g. `[d+]`.